### PR TITLE
feat(NodePorts): return Skip status in PreFilter

### LINKED ISSUE #114399: [Umbrella] Skip Filter plugins when coupled PreFilter plugin returns `Skip` status
### What would you like to be added?

It's spawned from https://github.com/kubernetes/kubernetes/issues/107556.

After https://github.com/kubernetes/kubernetes/pull/114125 gets merged, the scheduler skips the plugin in Filter, if that plugin return `Skip` in PreFilter. In other words, the plugin can return `Skip` when it doesn't do anything on Pod's scheduling.

For example, if the pod doesn't have any inter-pod affinity, the inter-pod affinity plugin PreFilter can return Skip, and the scheduler won't run the inter-pod affinity plugin Filter. 

Here, I just listed all filter plugins, **not all of them may not need to return Skip in PreFilter though.** Assignee, please make sure your plugin really needs to return `Skip` or not.
Please feel free to assign what you'd like to work on. And please comment in this issue before starting to work on so that we can avoid many people from working on the same thing!

- [x] InterPodAffinity @sanposhiho https://github.com/kubernetes/kubernetes/pull/114889
- [x] NodeAffinity @sanposhiho https://github.com/kubernetes/kubernetes/pull/114125
- [x] NodeName @1978629634 https://github.com/kubernetes/kubernetes/pull/114386
- [ ] NodePorts @tangwz  https://github.com/kubernetes/kubernetes/pull/118271
- [x] NodeResourceFit @tangwz (no need to implement)
- [x] NodeUnschedulable @sivchari (no need to implement)
- [x] CSILimits @tangwz https://github.com/kubernetes/kubernetes/pull/115398
- [x] nonCSILimits @tangwz https://github.com/kubernetes/kubernetes/pull/115398
- [ ] TopologySpread @utam0k https://github.com/kubernetes/kubernetes/pull/117683
- [x] TaintToleration (no need to implement) @sivchari https://github.com/kubernetes/kubernetes/pull/114405
- [x] VolumeBinding @kidddddddddddddddddddddd https://github.com/kubernetes/kubernetes/pull/114897
- [x] VolumeRestriction @kidddddddddddddddddddddd https://github.com/kubernetes/kubernetes/pull/114898
- [x] VolumeZone @kidddddddddddddddddddddd https://github.com/kubernetes/kubernetes/pull/114896

### Why is this needed?

The scheduler can know which plugins it really needs/doesn't need to run to get the scheduling result.
It'd contribute to the scheduling performance.

/sig scheduling
/assign

### PR DESCRIPTION
**Pull request by tangwz**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Update NodePorts PreFilter return Skip status if pod has no container ports on the NodePorts PreFilter.
This PR will improve the scheduling efficiency of pod.

#### Which issue(s) this PR fixes:

Fixes #114399

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/nodeports/node_ports_test.go [hunk 1]
```diff
@@ -47,15 +47,17 @@ func newPod(host string, hostPortInfos ...string) *v1.Pod {
...
+		wantPreFilterStatus *framework.Status
+		wantFilterStatus    *framework.Status
 	}{
 		{
-			pod:      &v1.Pod{},
-			nodeInfo: framework.NewNodeInfo(),
-			name:     "nothing running",
+			pod:                 &v1.Pod{},
+			nodeInfo:            framework.NewNodeInfo(),
+			name:                "skip scheduling",
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:59]:
```suggestion
			name:                "skip Filter",
```

File: pkg/scheduler/framework/plugins/nodeports/node_ports_test.go [hunk 2]
```diff
@@ -144,12 +146,18 @@ func TestNodePorts(t *testing.T) {
 			p, _ := New(nil, nil)
 			cycleState := framework.NewCycleState()
 			_, preFilterStatus := p.(framework.PreFilterPlugin).PreFilter(context.Background(), cycleState, test.pod)
+			if diff := cmp.Diff(test.wantPreFilterStatus, preFilterStatus); diff != "" {
+				t.Errorf("preFilter status does not match (-want,+got): %s", diff)
+			}
+			if preFilterStatus.IsSkip() {
+				return
+			}
 			if !preFilterStatus.IsSuccess() {
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:155]: I think we can remove this now.

**Review by kerthcet — COMMENTED** [review]
Some nits.

**Review by kerthcet — COMMENTED** [review]
Some nits.

File: pkg/scheduler/framework/plugins/nodeports/node_ports.go [hunk 3]
```diff
@@ -76,6 +76,10 @@ func getContainerPorts(pods ...*v1.Pod) []*v1.ContainerPort {
 // PreFilter invoked at the prefilter extension point.
 func (pl *NodePorts) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) (*framework.PreFilterResult, *framework.Status) {
 	s := getContainerPorts(pod)
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/nodeports/node_ports.go:78]: @tangwz 
Can you take a look at this suggestion from @/ahg-g?
https://github.com/kubernetes/kubernetes/pull/115318/files#r1169034360

File: pkg/scheduler/framework/plugins/nodeports/node_ports_test.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:155]: I think it's safer option to keep it for the future change even though the current PreFilter doesn't return error.

**Review by sanposhiho — COMMENTED** [review]
Thanks @tangwz for taking over this work!

**Review by sanposhiho — COMMENTED** [review]
Thanks @tangwz for taking over this work!

File: pkg/scheduler/framework/plugins/nodeports/node_ports.go (hunk 3 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/nodeports/node_ports.go:78]: Yes, lets optimize this further, we only care about host ports.

**Comment by tangwz** [comment]
> Thanks @tangwz for taking over this work!

According to the discussion, I just fix some typo.

File: pkg/scheduler/framework/plugins/nodeports/node_ports_test.go (hunk 2 above)
Comment (tangwz) [reply pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:155]: > I think it's safer option to keep it for the future change even though the current PreFilter doesn't return error.

agree to sanposhiho.

**Comment by ahg-g** [comment]
> @sanposhiho @kerthcet @ahg-g ping PTAL

Can you address https://github.com/kubernetes/kubernetes/pull/118271/files#r1206516333 pls

**Comment by tangwz** [comment]
> > @sanposhiho @kerthcet @ahg-g ping PTAL
> 
> Can you address https://github.com/kubernetes/kubernetes/pull/118271/files#r1206516333 pls

I was negligent, let me fix it.

File: pkg/scheduler/framework/plugins/nodeports/node_ports.go (hunk 3 above)
Comment (tangwz) [reply pkg/scheduler/framework/plugins/nodeports/node_ports.go:78]: > 

I think `CheckConflict(ip, protocol string, port int32)` need we return HostIP, Protocol and HostPort as the parameter, I don't understand how we only return host ports?

File: pkg/scheduler/framework/plugins/nodeports/node_ports_test.go [hunk 4]
```diff
@@ -212,48 +220,15 @@ func TestGetContainerPorts(t *testing.T) {
 					Protocol:      v1.ProtocolTCP,
 				},
 			}).Obj(),
-			expected: []*v1.ContainerPort{
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:215]: You should add the `HostPort` to all the cases above, and in addition, add a new container with no `HostPort` set.

Comment (ahg-g) [reply pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:215]: +1

**Comment by ahg-g** [comment]
thanks, pls squash

**Comment by tangwz** [comment]
> thanks, pls squash

@ahg-g @kerthcet  PTAL.

**Review by kerthcet — COMMENTED** [review]
/lgtm
/approve

Thanks @tangwz

**Pull request merged**